File: frontend/lib/borrowings-assets-contracts/progress.ts
/**
 * Section completion for Borrowings, Assets & Contracts.
 */

import { isFilledDecimal } from '@/lib/borrowings-assets-contracts/decimal';
import { BAC_CONFIRMATION_FIELDS } from '@/lib/borrowings-assets-contracts/options';
import type {
  BorrowingsAssetsContractsProgress,
  BorrowingsAssetsContractsSectionId,
  SectionStatus,
} from '@/lib/borrowings-assets-contracts/types';
import type { BorrowingsAssetsContractsPayload } from '@/lib/schemas/borrowings-assets-contracts';

function filled(value: unknown): boolean {
  if (value === null || value === undefined) return false;
  if (typeof value === 'string') return value.trim().length > 0;
  if (typeof value === 'boolean') return value;
  if (Array.isArray(value)) return value.length > 0;
  return true;
}

function statusFrom(answered: number, total: number, extraComplete = true): SectionStatus {
  if (answered === 0) return 'not_started';
  if (answered < total || !extraComplete) return 'in_progress';
  return 'complete';
}

export function evaluateFacilityMasterStatus(
  payload: BorrowingsAssetsContractsPayload,
): SectionStatus {
  const section = payload.financialIndebtednessAndFacilityMaster;
  const snapshot = section.borrowingSnapshot;
  const core = [
    filled(snapshot.positionAsOfDate),
    filled(snapshot.reportingCurrency),
    filled(snapshot.currentBorrowingsExist),
    section.facilities.length > 0,
  ];
  const answered = core.filter(Boolean).length;
  const facilitiesComplete = section.facilities.every(
    (facility) =>
      filled(facility.lender.lenderName) &&
      filled(facility.facilityType) &&
      (filled(facility.sanctionAndUtilisation.currentSanctionedLimit) ||
        isFilledDecimal(facility.sanctionAndUtilisation.principalOutstanding)),
  );
  return statusFrom(answered, core.length, facilitiesComplete);
}

export function evaluateSecurityChargesStatus(
  payload: BorrowingsAssetsContractsPayload,
): SectionStatus {
  const section = payload.securityChargesGuaranteesAndBorrowingPowers;
  const hasData =
    section.securities.length > 0 ||
    section.charges.length > 0 ||
    section.guarantees.length > 0 ||
    filled(section.borrowingPowers.boardBorrowingResolutionExists);
  if (!hasData) return 'not_started';

  const securitiesComplete = section.securities.every(
    (security) => filled(security.linkedFacilityId) && filled(security.securityType),
  );
  const chargesComplete = section.charges.every(
    (charge) => filled(charge.linkedFacilityId) && filled(charge.status),
  );
  return securitiesComplete && chargesComplete ? 'complete' : 'in_progress';
}

export function evaluateCovenantsConsentsStatus(
  payload: BorrowingsAssetsContractsPayload,
): SectionStatus {
  const section = payload.covenantsDefaultsWaiversAndLenderConsents;
  const hasData =
    section.covenants.length > 0 ||
    section.lenderConsents.length > 0 ||
    section.defaultEvents.length > 0;
  if (!hasData) return 'not_started';

  const covenantsComplete = section.covenants.every(
    (covenant) => filled(covenant.linkedFacilityId) && filled(covenant.covenantType),
  );
  const consentsComplete = section.lenderConsents.every(
    (consent) => filled(consent.linkedFacilityId) && filled(consent.ipoConsentRequirement),
  );
  return covenantsComplete && consentsComplete ? 'complete' : 'in_progress';
}

export function evaluatePropertiesStatus(
  payload: BorrowingsAssetsContractsPayload,
): SectionStatus {
  const section = payload.immovablePropertiesAndOccupancyRights;
  if (section.properties.length === 0) return 'not_started';
  const complete = section.properties.every(
    (property) =>
      filled(property.identity.propertyName) || filled(property.identity.address),
  );
  return complete ? 'complete' : 'in_progress';
}

export function evaluateAssetsStatus(payload: BorrowingsAssetsContractsPayload): SectionStatus {
  const section = payload.materialAssetsEncumbranceAndInsuranceLinkage;
  const hasData =
    section.assets.length > 0 ||
    section.insuranceLinkages.length > 0 ||
    section.ipContractualDependencies.length > 0;
  if (!hasData) return 'not_started';
  const complete = section.assets.every(
    (asset) => filled(asset.description) || filled(asset.assetClass),
  );
  return complete ? 'complete' : 'in_progress';
}

export function evaluateContractsStatus(
  payload: BorrowingsAssetsContractsPayload,
): SectionStatus {
  const contracts = payload.materialBusinessStrategicAndOtherContracts.contracts;
  if (contracts.length === 0) return 'not_started';
  const complete = contracts.every(
    (contract) =>
      filled(contract.basicTerms.agreementTitle) || filled(contract.parties.counterparty),
  );
  return complete ? 'complete' : 'in_progress';
}

export function evaluateMaterialityInspectionStatus(
  payload: BorrowingsAssetsContractsPayload,
): SectionStatus {
  const section = payload.contractMaterialityExpiryAndInspectionReadiness;
  const hasData =
    section.materialityRecords.length > 0 ||
    section.nonOrdinaryCourseReviews.length > 0 ||
    section.breachDisputeReadiness.length > 0 ||
    section.inspectionCandidates.length > 0;
  if (!hasData) return 'not_started';
  const complete = section.materialityRecords.every(
    (record) => filled(record.linkedContractId) && filled(record.materialityStatus),
  );
  return complete ? 'complete' : 'in_progress';
}

export function evaluateReconciliationConfirmationsStatus(
  payload: BorrowingsAssetsContractsPayload,
): SectionStatus {
  const confirmations = payload.reconciliationChangesAndIssuerConfirmations.confirmations;
  const answered = BAC_CONFIRMATION_FIELDS.filter((field) => confirmations[field.key] !== '').length;
  if (answered === 0) return 'not_started';
  if (answered < BAC_CONFIRMATION_FIELDS.length) return 'in_progress';
  return 'complete';
}

const SECTION_EVALUATORS: Record<
  BorrowingsAssetsContractsSectionId,
  (payload: BorrowingsAssetsContractsPayload) => SectionStatus
> = {
  'financial-indebtedness-and-facility-master': evaluateFacilityMasterStatus,
  'security-charges-guarantees-and-borrowing-powers': evaluateSecurityChargesStatus,
  'covenants-defaults-waivers-and-lender-consents': evaluateCovenantsConsentsStatus,
  'immovable-properties-and-occupancy-rights': evaluatePropertiesStatus,
  'material-assets-encumbrance-and-insurance-linkage': evaluateAssetsStatus,
  'material-business-strategic-and-other-contracts': evaluateContractsStatus,
  'contract-materiality-expiry-and-inspection-readiness': evaluateMaterialityInspectionStatus,
  'reconciliation-changes-and-issuer-confirmations': evaluateReconciliationConfirmationsStatus,
};

export function calculateBorrowingsAssetsContractsProgress(
  payload: BorrowingsAssetsContractsPayload,
): BorrowingsAssetsContractsProgress {
  const sections = Object.fromEntries(
    (Object.keys(SECTION_EVALUATORS) as BorrowingsAssetsContractsSectionId[]).map((sectionId) => [
      sectionId,
      SECTION_EVALUATORS[sectionId](payload),
    ]),
  ) as BorrowingsAssetsContractsProgress['sections'];

  const sectionsComplete = Object.values(sections).filter((status) => status === 'complete').length;
  const overallStatus: SectionStatus =
    sectionsComplete === 0
      ? 'not_started'
      : sectionsComplete === Object.keys(sections).length
        ? 'complete'
        : 'in_progress';

  return {
    sections,
    sectionsComplete,
    totalSections: Object.keys(sections).length,
    overallStatus,
  };
}

File: backend/app/modules/borrowings_assets_contracts/progress.py
"""Section completion for Borrowings, Assets & Contracts."""

from __future__ import annotations

from typing import Any, Callable

from app.modules.borrowings_assets_contracts import decimal_utils as dm
from app.modules.borrowings_assets_contracts.constants import (
    BAC_CONFIRMATION_FIELDS,
    SECTION_IDS,
)


def _filled(value: Any) -> bool:
    if value is None:
        return False
    if isinstance(value, str):
        return bool(value.strip())
    if isinstance(value, bool):
        return value
    if isinstance(value, list):
        return len(value) > 0
    return True


def _status_from(answered: int, total: int, extra_complete: bool = True) -> str:
    if answered == 0:
        return "not_started"
    if answered < total or not extra_complete:
        return "in_progress"
    return "complete"


def evaluate_facility_master_status(payload: dict[str, Any]) -> str:
    section = payload.get("financialIndebtednessAndFacilityMaster") or {}
    snapshot = section.get("borrowingSnapshot") or {}
    facilities = [f for f in (section.get("facilities") or []) if isinstance(f, dict)]
    core = [
        _filled(snapshot.get("positionAsOfDate")),
        _filled(snapshot.get("reportingCurrency")),
        _filled(snapshot.get("currentBorrowingsExist")),
        len(facilities) > 0,
    ]
    answered = sum(1 for item in core if item)
    facilities_complete = all(
        _filled((facility.get("lender") or {}).get("lenderName"))
        and _filled(facility.get("facilityType"))
        and (
            _filled((facility.get("sanctionAndUtilisation") or {}).get("currentSanctionedLimit"))
            or dm.is_filled(
                str((facility.get("sanctionAndUtilisation") or {}).get("principalOutstanding") or "")
            )
        )
        for facility in facilities
    )
    return _status_from(answered, len(core), facilities_complete)


def evaluate_security_charges_status(payload: dict[str, Any]) -> str:
    section = payload.get("securityChargesGuaranteesAndBorrowingPowers") or {}
    has_data = bool(
        section.get("securities")
        or section.get("charges")
        or section.get("guarantees")
        or _filled((section.get("borrowingPowers") or {}).get("boardBorrowingResolutionExists"))
    )
    if not has_data:
        return "not_started"

    securities = [s for s in (section.get("securities") or []) if isinstance(s, dict)]
    charges = [c for c in (section.get("charges") or []) if isinstance(c, dict)]
    securities_complete = all(
        _filled(security.get("linkedFacilityId")) and _filled(security.get("securityType"))
        for security in securities
    )
    charges_complete = all(
        _filled(charge.get("linkedFacilityId")) and _filled(charge.get("status")) for charge in charges
    )
    return "complete" if securities_complete and charges_complete else "in_progress"


def evaluate_covenants_consents_status(payload: dict[str, Any]) -> str:
    section = payload.get("covenantsDefaultsWaiversAndLenderConsents") or {}
    has_data = bool(
        section.get("covenants") or section.get("lenderConsents") or section.get("defaultEvents")
    )
    if not has_data:
        return "not_started"

    covenants = [c for c in (section.get("covenants") or []) if isinstance(c, dict)]
    consents = [c for c in (section.get("lenderConsents") or []) if isinstance(c, dict)]
    covenants_complete = all(
        _filled(covenant.get("linkedFacilityId")) and _filled(covenant.get("covenantType"))
        for covenant in covenants
    )
    consents_complete = all(
        _filled(consent.get("linkedFacilityId")) and _filled(consent.get("ipoConsentRequirement"))
        for consent in consents
    )
    return "complete" if covenants_complete and consents_complete else "in_progress"


def evaluate_properties_status(payload: dict[str, Any]) -> str:
    section = payload.get("immovablePropertiesAndOccupancyRights") or {}
    properties = [p for p in (section.get("properties") or []) if isinstance(p, dict)]
    if not properties:
        return "not_started"
    complete = all(
        _filled((property_.get("identity") or {}).get("propertyName"))
        or _filled((property_.get("identity") or {}).get("address"))
        for property_ in properties
    )
    return "complete" if complete else "in_progress"


def evaluate_assets_status(payload: dict[str, Any]) -> str:
    section = payload.get("materialAssetsEncumbranceAndInsuranceLinkage") or {}
    has_data = bool(
        section.get("assets")
        or section.get("insuranceLinkages")
        or section.get("ipContractualDependencies")
    )
    if not has_data:
        return "not_started"
    assets = [a for a in (section.get("assets") or []) if isinstance(a, dict)]
    complete = all(_filled(asset.get("description")) or _filled(asset.get("assetClass")) for asset in assets)
    return "complete" if complete else "in_progress"


def evaluate_contracts_status(payload: dict[str, Any]) -> str:
    contracts = [
        c
        for c in (
            (payload.get("materialBusinessStrategicAndOtherContracts") or {}).get("contracts") or []
        )
        if isinstance(c, dict)
    ]
    if not contracts:
        return "not_started"
    complete = all(
        _filled((contract.get("basicTerms") or {}).get("agreementTitle"))
        or _filled((contract.get("parties") or {}).get("counterparty"))
        for contract in contracts
    )
    return "complete" if complete else "in_progress"


def evaluate_materiality_inspection_status(payload: dict[str, Any]) -> str:
    section = payload.get("contractMaterialityExpiryAndInspectionReadiness") or {}
    has_data = bool(
        section.get("materialityRecords")
        or section.get("nonOrdinaryCourseReviews")
        or section.get("breachDisputeReadiness")
        or section.get("inspectionCandidates")
    )
    if not has_data:
        return "not_started"
    records = [r for r in (section.get("materialityRecords") or []) if isinstance(r, dict)]
    complete = all(
        _filled(record.get("linkedContractId")) and _filled(record.get("materialityStatus"))
        for record in records
    )
    return "complete" if complete else "in_progress"


def evaluate_reconciliation_confirmations_status(payload: dict[str, Any]) -> str:
    confirmations = (payload.get("reconciliationChangesAndIssuerConfirmations") or {}).get(
        "confirmations"
    ) or {}
    answered = sum(
        1 for key, _ in BAC_CONFIRMATION_FIELDS if confirmations.get(key) not in (None, "")
    )
    if answered == 0:
        return "not_started"
    if answered < len(BAC_CONFIRMATION_FIELDS):
        return "in_progress"
    return "complete"


SECTION_EVALUATORS: dict[str, Callable[[dict[str, Any]], str]] = {
    "financial-indebtedness-and-facility-master": evaluate_facility_master_status,
    "security-charges-guarantees-and-borrowing-powers": evaluate_security_charges_status,
    "covenants-defaults-waivers-and-lender-consents": evaluate_covenants_consents_status,
    "immovable-properties-and-occupancy-rights": evaluate_properties_status,
    "material-assets-encumbrance-and-insurance-linkage": evaluate_assets_status,
    "material-business-strategic-and-other-contracts": evaluate_contracts_status,
    "contract-materiality-expiry-and-inspection-readiness": evaluate_materiality_inspection_status,
    "reconciliation-changes-and-issuer-confirmations": evaluate_reconciliation_confirmations_status,
}


def calculate_borrowings_assets_contracts_progress(payload: dict[str, Any]) -> dict[str, Any]:
    sections = {section_id: SECTION_EVALUATORS[section_id](payload) for section_id in SECTION_IDS}
    sections_complete = sum(1 for status in sections.values() if status == "complete")
    total_sections = len(sections)
    if sections_complete == 0:
        overall_status = "not_started"
    elif sections_complete == total_sections:
        overall_status = "complete"
    else:
        overall_status = "in_progress"
    return {
        "sections": sections,
        "sectionsComplete": sections_complete,
        "totalSections": total_sections,
        "overallStatus": overall_status,
    }
